### ISSUE TITLE
amp-carousel: dynamically disabled previous/next buttons lose focus

### ISSUE THREAD

**Issue by TetraLogicalHelpdesk** [ampproject/amphtml#31169]
## WCAG Level
Level A

## Priority
Medium

## Type
Navigation

## Pages/screens/components affected
* [amp-base-carousel](https://amp.dev/documentation/components/amp-base-carousel/)
* [amp-stream-gallery](https://amp.dev/documentation/components/amp-stream-gallery/)
* [amp-inline-gallery](https://amp.dev/documentation/components/amp-inline-gallery/)
* [amp-lightbox-gallery](https://amp.dev/documentation/components/amp-lightbox-gallery/)

## Description
For people who navigate using a keyboard or keyboard-like device (such as a switch control or a joystick) the focus order of interactive controls in a document should be logical and intuitive. In particular, when a page changes dynamically as a result of user interaction, it's important that focus is, where necessary, moved to the most appropriate place within the page.

In non-looping carousels (such as the [`amp-base-carousel` looping Glitch](https://nutritious-tide-hornet.glitch.me/amp-base-carousel/grouping.html), [`amp-stream-gallery` basic Glitch](https://nutritious-tide-hornet.glitch.me/amp-stream-gallery/basic.html), the first usage example in the [`amp-inline-gallery` documentation](https://amp.dev/documentation/components/amp-inline-gallery/), and the "with `amp-carousel`" example in the [`amp-lightbox-gallery` documentation](https://amp.dev/documentation/examples/components/amp-lightbox-gallery/#with-amp-carousel)), the previous/next buttons are not displayed when the carousel is showing the first/last item, respectively.

In most cases, the `<button>` element used for the previous/next controls is dynamically set to `disabled`. In the case of the `amp-lightbox-gallery` example, the control is a custom button and disabled/hidden using CSS:

```
amp-carousel .amp-carousel-button.amp-disabled {
    animation: none;
    opacity: 0;
    visibility: hidden;
}
```

`<button disabled>` elements don't normally receive keyboard focus. In the case of the `amp-lightbox-gallery` example, the styling added to the element effectively removes it from the page - it is not keyboard-focusable anymore (despite the `tabindex="0"`), and hidden from assistive technologies/screen readers.

In both cases, this makes it problematic when a keyboard user operates the carousel/gallery controls, as the control that they're focused on will essentially "disappear" from under their focus, or become unfocusable. Depending on the specific browser/assistive technology used, this can result in the focus being lost and reset to the start of the page.

## User impact
When focus is not programmatically moved to the most appropriate location following an interaction, users will be left to manually navigate again to the most appropriate location in the page. For sighted keyboard users, this can be an inconvenience/irritation, particularly if focus was lost/reset to the start of the document and they have to cycle through the entire page again. Blind and low vision assistive technology users may be unaware of what change actually occurred on the page as a result of their interaction. After performing an action, these users may be confused/left wondering if their action was actually successful. In cases where focus is lost (generally caused by the element that previously had focus being disabled or removed from the page), some browsers may error-correct and attempt to relocate the user's focus to the nearest focusable element, or they may simply reset the focus to the start of the document. Overall, it will likely make the user's experience of navigating the page unnecessarily tedious and confusing.

## Required solution

For the `amp-lightbox-gallery` example, see the issue previously filed for [`amp-carousel`: Incorrect focus handling](https://github.com/ampproject/amphtml/issues/29558).

For the cases where the `<button>` element is dynamically set to be `disabled`, we'd recommend instead not setting this attribute, but instead adding `aria-disabled="true"` (which only signals to assistive technologies that the button is disabled), `tabindex="-1"` (to remove the control from regular keyboard focus order), and making sure the associated JavaScript does not execute when the button is set to be disabled this way.

```
<button ... aria-disabled="true" tabindex="-1">
  ...
</button>
```

As a result, the control will still retain focus when it is "disabled", preventing the focus from being lost/reset.

## Test procedure(s)
Use these steps to confirm that the solution has been correctly applied:
1. Use `Tab`/`Shift`+`Tab` to move keyboard focus to the previous/next buttons in a non-looping carousel.
2. Interact with the buttons until you reach the first or last slide.
3. Verify that when the previous or next button disappear, focus is not reset to the start of the page (this can be difficult to verify, as some browsers error-correct for these situations - the safest way to test this is to use developer tools and to check `document.activeElement` in the console - if it returns `<body>`, then focus was lost/reset).

## Related standards
* [WCAG 2.1 Success Criterion 2.4.3 Focus Order (Level A)](https://www.w3.org/TR/WCAG21/#focus-order)

## More information
* [W3C - Placing the interactive elements in an order that follows sequences and relationships within the content](https://www.w3.org/WAI/WCAG21/Techniques/general/G59)

## Test data
Test date: November 2020
